sdk/core/core-rest-pipeline/src/policies/retryPolicy.ts

```diff
@@ -0,0 +1,147 @@
+ */
+const retryPolicyName = "retryPolicy";
+
+/**
+ * Options to the {@link retryPolicy}
+ */
+export interface RetryPolicyOptions {
+  /**
+   * Maximum number of retries. If not specified, it will limit to 3 retries.
+   */
+  maxRetries: number;
+}
+
+/**
+ * retryPolicy is a generic policy to enable retrying requests when certain conditions are met
+ */
+export function retryPolicy(
+  strategies: RetryStrategy[],
+  options: RetryPolicyOptions = { maxRetries: DEFAULT_MAX_RETRIES }
+): PipelinePolicy {
+  return {
+    name: retryPolicyName,
+    async sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse> {
+      let response: PipelineResponse | undefined;
+      let responseError: RestError | undefined;
+      let retryCount = -1;
+
+      // eslint-disable-next-line no-constant-condition
+      retryRequest: while (true) {
+        retryCount += 1;
+        response = undefined;
+        responseError = undefined;
+
+        try {
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Attempting to send request`,
+            request.requestId
+          );
+          response = await next(request);
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Received a response from request`,
+            request.requestId
+          );
+        } catch (e) {
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Received an error from request`,
+            request.requestId
+          );
+          responseError = e as RestError;
+          if (responseError.name !== "RestError") {
+            throw responseError;
+          }
+          response = responseError.response;
+        }
+
+        retryPolicyLogger.info(
+          `Retry ${retryCount}: Processing ${strategies.length} retry strategies.`
+        );
+
+        strategiesLoop: for (const strategy of strategies) {
+          const strategyLogger = strategy.logger || retryPolicyLogger;
+          strategyLogger.info(`Retry ${retryCount}: Processing retry strategy ${strategy.name}.`);
+
+          const information: RetryInformation = {
+            retryCount,
+            response,
+            responseError
+          };
+
+          if (information.retryCount >= options.maxRetries ?? DEFAULT_MAX_RETRIES) {
+            strategyLogger.info(
+              `Maximum retries reached. Returning the last received response, or throwing the last received error.`
+            );
+            if (response !== undefined) {
+              return response;
+            }
+            throw responseError;
+          }
+
+          let modifiers: RetryModifiers;
+          try {
+            modifiers = strategy.retry(information);
+          } catch (e) {
+            const error = e as Error;
+            if (error.name === "SkipRetryError") {
+              strategyLogger.error(`Retry ${retryCount}: Skipped. ${error.message}`);
+              continue strategiesLoop;
+            }
+            throw e;
+          }
```

Comment:
I'm not a fan of exceptions for control flow. Why not just add a modifier for this and early return from the strategy?

```suggestion
          modifiers = strategy.retry(information);
          if (modifiers.skipStrategy) {
             strategyLogger.error(`Retry ${retryCount}: Skipped.`);
             continue strategiesLoop;
          }
```

Reply:
Oh ok ok! I thought the error would be more similar to how Identity handles UnavailableCredential and such. Familiarity could be a plus!

Reply:
Moved to the skipStrategy modifier approach here: https://github.com/Azure/azure-sdk-for-js/pull/19078/commits/7f84291de33c9137f9c0257251815095c004b9e1